File: src/main/java/com/king/java/field/plugin/entity/FieldParseConfig.java
/*
 * Copyright (C) 2020 Jenly Yu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.king.java.field.plugin.entity;

import com.king.java.field.plugin.dialog.GenerateFieldDialog;
import org.jetbrains.annotations.NotNull;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * 字段解析配置
 *
 * @author <a href="mailto:yujinlin@mail.com">Jenly</a>
 */
public class FieldParseConfig {

    /**
     * 字段所在列；默认：第0列
     */
    private int fieldColumn = 0;
    /**
     * 字段类型所在列；默认：第1列
     */
    private int fieldTypeColumn = 1;
    /**
     * 字段注释说明所在列；默认：第2列
     */
    private int fieldCommentColumn = 2;
    /**
     * 字段修饰符；默认：private
     */
    private String fieldModifier = GenerateFieldDialog.Modifier.PRIVATE.getValue();
    /**
     * 是否生成 get 和 set 方法
     */
    private boolean generateGetterAndSetter;
    /**
     * 是否生成 toString 方法
     */
    private boolean generateToString;
    /**
     * 字段类型转换配置；key为原始类型，value表示需要目标类型；如：key = int，value = Integer 时，则会在生成字段类型时，自动将 int 转换为 Integer
     */
    private Map<String,String> filedTypeConvertMap;
    /**
     * 是否将数组转换为列表
     */
    private boolean convertArrayToList = true;

    public int getFieldColumn() {
        return fieldColumn;
    }

    public void setFieldColumn(int fieldColumn) {
        this.fieldColumn = fieldColumn;
    }

    public int getFieldTypeColumn() {
        return fieldTypeColumn;
    }

    public void setFieldTypeColumn(int fieldTypeColumn) {
        this.fieldTypeColumn = fieldTypeColumn;
    }

    public int getFieldCommentColumn() {
        return fieldCommentColumn;
    }

    public void setFieldCommentColumn(int fieldCommentColumn) {
        this.fieldCommentColumn = fieldCommentColumn;
    }

    public String getFieldModifier() {
        return fieldModifier;
    }

    public void setFieldModifier(String fieldModifier) {
        this.fieldModifier = fieldModifier;
    }

    public boolean isGenerateToString() {
        return generateToString;
    }

    public void setGenerateToString(boolean generateToString) {
        this.generateToString = generateToString;
    }

    public boolean isGenerateGetterAndSetter() {
        return generateGetterAndSetter;
    }

    public void setGenerateGetterAndSetter(boolean generateGetterAndSetter) {
        this.generateGetterAndSetter = generateGetterAndSetter;
    }

    @NotNull
    public Map<String, String> getFiledTypeConvertMap() {
        if (filedTypeConvertMap == null) {
            filedTypeConvertMap = createFiledTypeConvertMap();
        }
        return filedTypeConvertMap;
    }

    private Map<String, String> createFiledTypeConvertMap(){
        Map<String, String> filedTypeConvertMap = new LinkedHashMap<>(16);
        filedTypeConvertMap.put("varchar", "String");
        filedTypeConvertMap.put("tinytext", "String");
        filedTypeConvertMap.put("text", "String");
        filedTypeConvertMap.put("string", "String");
        filedTypeConvertMap.put("bool", "boolean");
        filedTypeConvertMap.put("integer", "Integer");
        return filedTypeConvertMap;
    }

    public void setFiledTypeConvertMap(Map<String, String> filedTypeConvertMap) {
        this.filedTypeConvertMap = filedTypeConvertMap;
    }

    public boolean isConvertArrayToList() {
        return convertArrayToList;
    }

    public void setConvertArrayToList(boolean convertArrayToList) {
        this.convertArrayToList = convertArrayToList;
    }

    @Override
    public String toString() {
        return "FieldParseConfig{" +
                "fieldColumn=" + fieldColumn +
                ", fieldTypeColumn=" + fieldTypeColumn +
                ", fieldCommentColumn=" + fieldCommentColumn +
                ", fieldModifier='" + fieldModifier + '\'' +
                ", generateGetterAndSetter=" + generateGetterAndSetter +
                ", generateToString=" + generateToString +
                ", filedTypeConvertMap=" + filedTypeConvertMap +
                ", convertArrayToList=" + convertArrayToList +
                '}';
    }
}

File: src/main/java/com/king/jvm/field/generator/model/FieldParseConfig.kt
/*
 * Copyright (C) 2020 Jenly Yu
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.king.jvm.field.generator.model

import java.util.LinkedHashMap

/**
 * 字段解析配置模型
 *
 * @author <a href="mailto:jenly1314@gmail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
class FieldParseConfig {

    var fieldColumn: Int = DEFAULT_FIELD_COLUMN
    var fieldTypeColumn: Int = DEFAULT_FIELD_TYPE_COLUMN
    var fieldCommentColumn: Int = DEFAULT_FIELD_COMMENT_COLUMN
    var fieldNotNullColumn: Int = DEFAULT_FIELD_NOTNULL_COLUMN
    var valueSeparator: ValueSeparator = ValueSeparator.TAB
    var fieldNameStyle: FieldNameStyle = FieldNameStyle.NONE
    var fieldSortStyle: FieldSortStyle = FieldSortStyle.DEFAULT
    var existingFieldPolicy: ExistingFieldPolicy = ExistingFieldPolicy.IGNORE_NEW
    var notNullKeywords: String = DEFAULT_NOT_NULL_KEYWORDS
    var nullabilityMode: NullabilityMode = NullabilityMode.AUTO
    var fieldModifier: Modifier = Modifier.PRIVATE
    var targetLanguage: TargetLanguage = TargetLanguage.AUTO
    var isLastGeneratedKotlin: Boolean? = null
    var isEnableInputRegexValidation: Boolean = true
    var isGenerateGetterAndSetter: Boolean = true
    var isGenerateToString: Boolean = true
    var isConvertArrayToList: Boolean = true
    var isUseDataClass: Boolean = true
    var kotlinPropertyKeyword: KotlinPropertyKeyword = KotlinPropertyKeyword.VAL

    var annotationType: AnnotationType = AnnotationType.NONE
    var customAnnotationImport: String = ""
    var customClassAnnotation: String = ""
    var customPropertyAnnotation: String = ""

    var fieldTypeConvertMap: MutableMap<String, String?>? = null
        get() {
            if (field == null) {
                field = createFieldTypeConvertMap()
            }
            return field
        }

    private fun createFieldTypeConvertMap(): MutableMap<String, String?> {
        val convertMap = LinkedHashMap<String, String?>(48)
        convertMap["char"] = "String"
        convertMap["nchar"] = "String"
        convertMap["varchar"] = "String"
        convertMap["varchar2"] = "String"
        convertMap["nvarchar"] = "String"
        convertMap["tinytext"] = "String"
        convertMap["text"] = "String"
        convertMap["mediumtext"] = "String"
        convertMap["longtext"] = "String"
        convertMap["ntext"] = "String"
        convertMap["json"] = "String"
        convertMap["jsonb"] = "String"
        convertMap["uuid"] = "String"
        convertMap["string"] = "String"
        convertMap["bool"] = "boolean"
        convertMap["boolean"] = "boolean"
        convertMap["bit"] = "boolean"
        convertMap["tinyint"] = "Byte"
        convertMap["smallint"] = "Short"
        convertMap["int"] = "Integer"
        convertMap["integer"] = "Integer"
        convertMap["int4"] = "Integer"
        convertMap["bigint"] = "Long"
        convertMap["int8"] = "Long"
        convertMap["float"] = "Float"
        convertMap["float4"] = "Float"
        convertMap["double"] = "Double"
        convertMap["float8"] = "Double"
        convertMap["decimal"] = "Double"
        convertMap["numeric"] = "Double"
        convertMap["number"] = "Double"
        convertMap["real"] = "Double"
        convertMap["date"] = "String"
        convertMap["datetime"] = "String"
        convertMap["timestamp"] = "String"
        convertMap["time"] = "String"
        convertMap["year"] = "Integer"
        return convertMap
    }

    fun ensureDefaults(): FieldParseConfig {
        valueSeparator = valueOrDefault(valueSeparator, ValueSeparator.TAB)
        fieldNameStyle = valueOrDefault(fieldNameStyle, FieldNameStyle.NONE)
        fieldSortStyle = valueOrDefault(fieldSortStyle, FieldSortStyle.DEFAULT)
        existingFieldPolicy = valueOrDefault(existingFieldPolicy, ExistingFieldPolicy.IGNORE_NEW)
        notNullKeywords = valueOrDefault(notNullKeywords, DEFAULT_NOT_NULL_KEYWORDS)
        nullabilityMode = valueOrDefault(nullabilityMode, NullabilityMode.AUTO)
        fieldModifier = valueOrDefault(fieldModifier, Modifier.PRIVATE)
        targetLanguage = valueOrDefault(targetLanguage, TargetLanguage.AUTO)
        kotlinPropertyKeyword = valueOrDefault(kotlinPropertyKeyword, KotlinPropertyKeyword.VAL)
        annotationType = valueOrDefault(annotationType, AnnotationType.NONE)
        customAnnotationImport = valueOrDefault(customAnnotationImport, "")
        customClassAnnotation = valueOrDefault(customClassAnnotation, "")
        customPropertyAnnotation = valueOrDefault(customPropertyAnnotation, "")
        if (fieldTypeConvertMap == null) {
            fieldTypeConvertMap = createFieldTypeConvertMap()
        }
        return this
    }

    private fun <T> valueOrDefault(value: T?, defaultValue: T): T {
        return value ?: defaultValue
    }

    companion object {
        const val DEFAULT_FIELD_COLUMN: Int = 0
        const val DEFAULT_FIELD_TYPE_COLUMN: Int = 1
        const val DEFAULT_FIELD_COMMENT_COLUMN: Int = -1
        const val DEFAULT_FIELD_NOTNULL_COLUMN: Int = 2
        const val DEFAULT_FIELD_TYPE: String = "Object"
        const val DEFAULT_NOT_NULL_KEYWORDS: String = "Y,是,必须"
    }
}
